flocker/testtools/test/test_base.py

```diff
@@ -0,0 +1,193 @@
+"""
+
+import errno
+import os
+import shutil
+import string
+
+from hypothesis import assume, given
+from hypothesis.strategies import binary, integers, lists, text
+from testtools import PlaceHolder, TestCase
+from testtools.matchers import (
+    AllMatch,
+    AfterPreprocessing,
+    Annotate,
+    Contains,
+    DirExists,
+    HasLength,
+    Equals,
+    FileContains,
+    Matcher,
+    MatchesAny,
+    LessThan,
+    Not,
+    PathExists,
+    StartsWith,
+)
+from testtools.testresult.doubles import Python27TestResult
+from twisted.python.filepath import FilePath
+from twisted.trial import unittest
+
+from .._base import (
+    AsyncTestCase,
+    make_temporary_directory,
+    _path_for_test_id,
+)
+
+
+class AsyncTestCaseTests(TestCase):
+    """
+    Tests for `AsyncTestCase`.
+    """
+
+    @given(binary(average_size=30))
+    def test_trial_skip_exception(self, reason):
+        """
+        If tests raise the ``SkipTest`` exported by Trial, then that's
+        recorded as a skip.
+        """
+
+        class SkippingTest(AsyncTestCase):
+            def test_skip(self):
+                raise unittest.SkipTest(reason)
+
+        test = SkippingTest('test_skip')
+        # We need a test result double that we can useful results from, and
+        # the Python 2.7 TestResult is the lowest common denominator.
+        result = Python27TestResult()
+        test.run(result)
+        # testing-cabal/testtools c51fdb854 adds a public API for this. Update
+        # to use new API when we start using a version later than 1.8.0.
+        self.assertEqual([
+            ('startTest', test),
+            ('addSkip', test, reason),
+            ('stopTest', test),
+        ], result._events)
+
+    def test_mktemp_doesnt_exist(self):
+        """
+        ``mktemp`` returns a path that doesn't exist inside a directory that
+        does.
+        """
+
+        class SomeTest(AsyncTestCase):
+            def test_pass(self):
+                pass
+
+        test = SomeTest('test_pass')
+        temp_path = FilePath(test.mktemp())
+        self.addCleanup(_remove_dir, temp_path.parent())
+
+        self.expectThat(temp_path.parent().path, DirExists())
+        self.expectThat(temp_path.path, Not(PathExists()))
+        self.assertThat(temp_path, BelowPath(FilePath(os.getcwd())))
+
+    def test_mktemp_not_deleted(self):
+        """
+        ``mktemp`` returns a path that's not deleted after the test is run.
+        """
+        created_files = []
+
+        class SomeTest(AsyncTestCase):
+            def test_create_file(self):
+                path = self.mktemp()
+                created_files.append(path)
+                open(path, 'w').write('hello')
+
+        SomeTest('test_create_file').run(Python27TestResult())
+        [path] = created_files
+        self.addCleanup(os.unlink, path)
+        self.assertThat(path, FileContains('hello'))
+
+
+identifier_characters = string.ascii_letters + string.digits + '_'
```

Comment:
Oh but they're ... Python identifier strings ...

nevermind, way too much of a mess.  Please disregard this.